presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestJoinNodeFlattener.java

```diff
@@ -0,0 +1,324 @@
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.LEFT;
+import static com.facebook.presto.sql.tree.ArithmeticBinaryExpression.Type.ADD;
+import static com.facebook.presto.sql.tree.ComparisonExpressionType.EQUAL;
+import static com.facebook.presto.sql.tree.ComparisonExpressionType.GREATER_THAN;
+import static com.facebook.presto.sql.tree.ComparisonExpressionType.LESS_THAN;
+import static com.facebook.presto.sql.tree.ComparisonExpressionType.NOT_EQUAL;
+import static com.facebook.presto.testing.TestingSession.testSessionBuilder;
+import static org.testng.Assert.assertEquals;
+
+public class TestJoinNodeFlattener
+{
+    private static final int DEFAULT_JOIN_LIMIT = 10;
+
+    private LocalQueryRunner queryRunner;
+
+    @BeforeClass
+    public void setUp()
+    {
+        queryRunner = new LocalQueryRunner(testSessionBuilder().build());
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void tearDown()
+    {
+        queryRunner.close();
+        queryRunner = null;
+    }
+
+    @Test(expectedExceptions = IllegalStateException.class)
+    public void testDoesNotFireForOuterJoins()
+    {
+        PlanBuilder p = new PlanBuilder(new PlanNodeIdAllocator(), queryRunner.getMetadata());
+        JoinNode outerJoin = p.join(
+                JoinNode.Type.FULL,
+                p.values(p.symbol("A1", BIGINT)),
+                p.values(p.symbol("B1", BIGINT)),
+                ImmutableList.of(new JoinNode.EquiJoinClause(p.symbol("A1", BIGINT), p.symbol("B1", BIGINT))),
+                ImmutableList.of(p.symbol("A1", BIGINT), p.symbol("B1", BIGINT)),
+                Optional.empty());
+        toMultiJoinNode(outerJoin, noLookup(), DEFAULT_JOIN_LIMIT);
+    }
+
+    @Test
+    public void testDoesNotConvertNestedOuterJoins()
+    {
+        PlanNodeIdAllocator idAllocator = new PlanNodeIdAllocator();
+        PlanBuilder planBuilder = new PlanBuilder(idAllocator, queryRunner.getMetadata());
+        Symbol a1 = planBuilder.symbol("A1", BIGINT);
+        Symbol b1 = planBuilder.symbol("B1", BIGINT);
+        Symbol c1 = planBuilder.symbol("C1", BIGINT);
+        JoinNode leftJoin = planBuilder.join(
+                LEFT,
+                planBuilder.values(a1),
+                planBuilder.values(b1),
+                ImmutableList.of(new JoinNode.EquiJoinClause(a1, b1)),
+                ImmutableList.of(a1, b1),
+                Optional.empty());
+        ValuesNode valuesC = planBuilder.values(c1);
+        JoinNode joinNode = planBuilder.join(
+                INNER,
+                leftJoin,
+                valuesC,
+                ImmutableList.of(new JoinNode.EquiJoinClause(a1, c1)),
+                ImmutableList.of(a1, b1, c1),
+                Optional.empty());
+
+        MultiJoinNode expected = new MultiJoinNode(new LinkedHashSet<>(ImmutableList.of(leftJoin, valuesC)), new ComparisonExpression(EQUAL, a1.toSymbolReference(), c1.toSymbolReference()), ImmutableList.of(a1, b1, c1));
```

Comment:
It is hard to see what's going on with all that boiler plate of `ImmutableList.of`, `new ComparisonExpression`, and `.toSymbolReference()`. Consider, adding a `MultiJoinNodeBuilder` like so:

```
        MultiJoinNode expected = new MultiJoinNodeBuilder()
                .withSources(leftJoin, valuesC)
                .withFilter(equals(a1, c1))
                .withOutputSymbols(a1, b1, c1)
                .build();

    private ComparisonExpression equals(Symbol left, Symbol right)
    {
        return new ComparisonExpression(EQUAL, left.toSymbolReference(), right.toSymbolReference());
    }

    private static final class MultiJoinNodeBuilder
    {
        private List<PlanNode> sources;
        private Expression filter;
        private List<Symbol> outputSymbols;

        public MultiJoinNodeBuilder withSources(PlanNode... sources)
        {
            this.sources = ImmutableList.copyOf(sources);
            return this;
        }

        public MultiJoinNodeBuilder withFilter(Expression filter)
        {
            this.filter = filter;
            return this;
        }

        public MultiJoinNodeBuilder withOutputSymbols(Symbol... outputSymbols)
        {
            this.outputSymbols = ImmutableList.copyOf(outputSymbols);
            return this;
        }

        public MultiJoinNode build()
        {
            return new MultiJoinNode(new LinkedHashSet<>(sources), filter, outputSymbols);
        }
    }
```